In [1]:
# importing libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# loading dataset
df = pd.read_csv(r"C:\Users\rasika\Downloads\sonardataset.csv")
df.head()

,x_1,x_2,x_3,x_4,x_5,x_6,x_7,x_8,x_9,x_10,...,x_52,x_53,x_54,x_55,x_56,x_57,x_58,x_59,x_60,Y
0,0.0200,0.0371,0.0428,0.0207,0.0954,0.0986,0.1539,0.1601,0.3109,0.2111,...,0.0027,0.0065,0.0159,0.0072,0.0167,0.0180,0.0084,0.0090,0.0032,R
1,0.0453,0.0523,0.0843,0.0689,0.1183,0.2583,0.2156,0.3481,0.3337,0.2872,...,0.0084,0.0089,0.0048,0.0094,0.0191,0.0140,0.0049,0.0052,0.0044,R
2,0.0262,0.0582,0.1099,0.1083,0.0974,0.2280,0.2431,0.3771,0.5598,0.6194,...,0.0232,0.0166,0.0095,0.0180,0.0244,0.0316,0.0164,0.0095,0.0078,R
3,0.0100,0.0171,0.0623,0.0205,0.0205,0.0368,0.1098,0.1276,0.0598,0.1264,...,0.0121,0.0036,0.0150,0.0085,0.0073,0.0050,0.0044,0.0040,0.0117,R
4,0.0762,0.0666,0.0481,0.0394,0.0590,0.0649,0.1209,0.2467,0.3564,0.4459,...,0.0031,0.0054,0.0105,0.0110,0.0015,0.0072,0.0048,0.0107,0.0094,R


In [3]:
# Dataset Summary
df.info() # number of data entries, number of columns and there datatype

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 208 entries, 0 to 207
Data columns (total 61 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   x_1     208 non-null    float64
 1   x_2     208 non-null    float64
 2   x_3     208 non-null    float64
 3   x_4     208 non-null    float64
 4   x_5     208 non-null    float64
 5   x_6     208 non-null    float64
 6   x_7     208 non-null    float64
 7   x_8     208 non-null    float64
 8   x_9     208 non-null    float64
 9   x_10    208 non-null    float64
 10  x_11    208 non-null    float64
 11  x_12    208 non-null    float64
 12  x_13    208 non-null    float64
 13  x_14    208 non-null    float64
 14  x_15    208 non-null    float64
 15  x_16    208 non-null    float64
 16  x_17    208 non-null    float64
 17  x_18    208 non-null    float64
 18  x_19    208 non-null    float64
 19  x_20    208 non-null    float64
 20  x_21    208 non-null    float64
 21  x_22    208 non-null    float64
 22  x_

In [4]:
# statistical summary
df.describe() # count, mean, standard deviation, min, max, interquartiles of data in features

,x_1,x_2,x_3,x_4,x_5,x_6,x_7,x_8,x_9,x_10,...,x_51,x_52,x_53,x_54,x_55,x_56,x_57,x_58,x_59,x_60
count,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,...,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000
mean,0.029164,0.038437,0.043832,0.053892,0.075202,0.104570,0.121747,0.134799,0.178003,0.208259,...,0.016069,0.013420,0.010709,0.010941,0.009290,0.008222,0.007820,0.007949,0.007941,0.006507
std,0.022991,0.032960,0.038428,0.046528,0.055552,0.059105,0.061788,0.085152,0.118387,0.134416,...,0.012008,0.009634,0.007060,0.007301,0.007088,0.005736,0.005785,0.006470,0.006181,0.005031
min,0.001500,0.000600,0.001500,0.005800,0.006700,0.010200,0.003300,0.005500,0.007500,0.011300,...,0.000000,0.000800,0.000500,0.001000,0.000600,0.000400,0.000300,0.000300,0.000100,0.000600
25%,0.013350,0.016450,0.018950,0.024375,0.038050,0.067025,0.080900,0.080425,0.097025,0.111275,...,0.008425,0.007275,0.005075,0.005375,0.004150,0.004400,0.003700,0.003600,0.003675,0.003100
50%,0.022800,0.030800,0.034300,0.044050,0.062500,0.092150,0.106950,0.112100,0.152250,0.182400,...,0.013900,0.011400,0.009550,0.009300,0.007500,0.006850,0.005950,0.005800,0.006400,0.005300
75%,0.035550,0.047950,0.057950,0.064500,0.100275,0.134125,0.154000,0.169600,0.233425,0.268700,...,0.020825,0.016725,0.014900,0.014500,0.012100,0.010575,0.010425,0.010350,0.010325,0.008525
max,0.137100,0.233900,0.305900,0.426400,0.401000,0.382300,0.372900,0.459000,0.682800,0.710600,...,0.100400,0.070900,0.039000,0.035200,0.044700,0.039400,0.035500,0.044000,0.036400,0.043900


In [5]:
# Checking for missing values
df.isnull().sum()

x_1     0
x_2     0
x_3     0
x_4     0
x_5     0
       ..
x_57    0
x_58    0
x_59    0
x_60    0
Y       0
Length: 61, dtype: int64

In [6]:
# checking for duplicates
df.duplicated().sum()

np.int64(0)

In [7]:
# Feature encoding on target column 'Y'
y = df['Y']

from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

y_le = le.fit_transform(y)

In [8]:
# Feature scaling to adjust features in one range
X = df.drop('Y', axis = 1)
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_sc = sc.fit_transform(X)

In [9]:
# Train test split with 80% data for training
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_sc, y_le, test_size = 0.2, random_state = 42, stratify = y)

### BUilding ANN model

In [10]:
!pip install tensorflow

In [11]:
from tensorflow.keras.models import Sequential # Creates neural network where layers are added one after another
from tensorflow.keras.layers import Dense # creates fully connected layer

model = Sequential() # creates empty neura network

model.add(Dense(32, activation = 'sigmoid', input_shape = (60, ))) # Creates hidden layer with 32 neurons, uses sigmoid function cause it is binary classification type of dataset and takes all 60 input features
model.add(Dense(16, activation = 'sigmoid')) # creates second hidden layer with 16 neurons and forms more complex pattern
model.add(Dense(1, activation = 'sigmoid')) # creates output layer with sigmoid that will range output between 0 and 1

C:\Users\rasika\anaconda3\Lib\site-packages\keras\src\layers\core\dense.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


### Compiling model

In [12]:
model.compile(optimizer = 'adam',
             loss = 'binary_crossentropy',
             metrics = ['accuracy']) 
# tells tensorflow how neural network should run.
# optimizer controls the weight updates and biases to reduce the errors
# loss tells how wrong predictions are 
# metrics to evalaute performance

### Training the model

In [13]:
history = model.fit(X_train, y_train, epochs = 100, batch_size = 16, validation_split = 0.2, verbose = 1)

Epoch 1/100
9/9 ━━━━━━━━━━━━━━━━━━━━ 5s 128ms/step - accuracy: 0.5758 - loss: 0.6884 - val_accuracy: 0.3529 - val_loss: 0.7340
Epoch 2/100
9/9 ━━━━━━━━━━━━━━━━━━━━ 0s 33ms/step - accuracy: 0.5758 - loss: 0.6739 - val_accuracy: 0.3824 - val_loss: 0.7327
Epoch 3/100
9/9 ━━━━━━━━━━━━━━━━━━━━ 0s 30ms/step - accuracy: 0.5758 - loss: 0.6592 - val_accuracy: 0.4118 - val_loss: 0.7191
Epoch 4/100
9/9 ━━━━━━━━━━━━━━━━━━━━ 0s 29ms/step - accuracy: 0.5833 - loss: 0.6478 - val_accuracy: 0.4118 - val_loss: 0.6999
Epoch 5/100
9/9 ━━━━━━━━━━━━━━━━━━━━ 0s 41ms/step - accuracy: 0.6061 - loss: 0.6354 - val_accuracy: 0.4706 - val_loss: 0.6896
Epoch 6/100
9/9 ━━━━━━━━━━━━━━━━━━━━ 1s 31ms/step - accuracy: 0.6364 - loss: 0.6235 - val_accuracy: 0.5882 - val_loss: 0.6751
Epoch 7/100
9/9 ━━━━━━━━━━━━━━━━━━━━ 0s 30ms/step - accuracy: 0.6818 - loss: 0.6125 - val_accuracy: 0.6176 - val_loss: 0.6643
Epoch 8/100
9/9 ━━━━━━━━━━━━━━━━━━━━ 0s 30ms/step - accuracy: 0.7045 - loss: 0.6000 - val_accuracy: 0.6471 - val_loss

### Predictions

In [14]:
y_prob = model.predict(X_test)
y_pred = (y_prob > 0.5).astype(int)

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 181ms/step


### Model Evaluation

In [26]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

print("Accuracy Score ", accuracy)
print("precision Score ", precision)
print("Recall Score ", recall)
print("F1 Score ", f1)

Accuracy Score  0.8095238095238095
precision Score  0.9285714285714286
Recall Score  0.65
F1 Score  0.7647058823529411


### HpyerParametr Tunning using grid search 

In [16]:
import sys
print(sys.executable)

C:\Users\rasika\anaconda3\python.exe


In [17]:
!pip install scikeras

In [18]:
from scikeras.wrappers import KerasClassifier
from sklearn.model_selection import GridSearchCV

In [19]:
def create_model(neurons=32):

    model = Sequential()

    model.add(Dense(neurons,
                    activation='relu',
                    input_shape=(60,)))

    model.add(Dense(1,
                    activation='sigmoid'))

    model.compile(
        optimizer='adam',
        loss='binary_crossentropy',
        metrics=['accuracy']
    )

    return model

In [20]:
model = KerasClassifier(
    model=create_model,
    verbose=0
)

In [ ]:
param_grid = {
    "model__neurons":[64],
    "batch_size":[32],
    "epochs":[100]
}

grid = GridSearchCV(
    estimator=model,
    param_grid=param_grid,
    cv=5,
    scoring='accuracy'
)

grid.fit(X_train, y_train)

In [24]:
# Best parameters and score
print(grid.best_params_)
print(grid.best_score_) 

{'batch_size': 32, 'epochs': 100, 'model__neurons': 64}
0.8433155080213904


### Metrics Evaluation on tunned model

In [29]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
#getting best model
best_model = grid.best_estimator_

# getting prediction with best model
y_prob_best = best_model.predict(X_test)
y_pred_best = (y_prob_best > 0.5).astype(int)

# Best Model evaluation
accuracy_grid = accuracy_score(y_test, y_pred_best)
precision_grid = precision_score(y_test, y_pred_best)
recall_grid = recall_score(y_test, y_pred_best)
f1_grid = f1_score(y_test, y_pred_best)

print("Accuracy Score ", accuracy_grid)
print("precision Score ", precision_grid)
print("Recall Score ", recall_grid)
print("F1 Score ", f1_grid)

Accuracy Score  0.8809523809523809
precision Score  0.9411764705882353
Recall Score  0.8
F1 Score  0.8648648648648649


###### Couldn't do parameter tunning with many combination at once cause parameter tunning is sensitive.
###### so tried with one combination only as one done with actual model

### Comparision and analysis

In [30]:
print('Model accuracy ', accuracy)
print('Model Accuracy after Parameter Tunning ', accuracy_grid)

Model accuracy  0.8095238095238095
Model Accuracy after Parameter Tunning  0.8809523809523809


With epochs 100, batch_size 16 and neuron 32 the model accuracy was 80% and after parameter tunning with 100 epochs, 64 neurons and 32 batch_size model accuracy is improved with 88%.
So it shows ANN can effectively classify sonar signal are whether mine or rocks

### Interpretation and Conclusion

1. The dataset contains 208 sonar observations with 60 numerical features and one binary target variable.
2. In dataset there is no missing or duplicates values present
3. Labels were encoded using label encoding and features standardized using StandardScaler to scale features around mean 0 and standard deviation 1.
4. A feedforward ANN with ReLU hidden layers and Sigmoid output layer was implemented.
5. The model was trained using Adam optimizer and Binary Crossentropy loss.
6. Hyperparameter tuning was performed by varying hidden layers, neurons, batch size, epochs, activation functions, and learning rate using Grid Search
7. The tuned model achieved higher accuracy, precision, recall, and F1-score than the default model.
8. Results indicate that ANN can effectively learn sonar signal patterns and distinguish mines from rocks.